## Exploring Ebay Car Sales Data


In this guided project, we'll work with a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website. This dataset cointains over 370,000 used cars scraped from Ebay Kleinanzeigen. The whole dataset can be downloaded from [here](https://www.kaggle.com/orgesleka/used-cars-database/data).

Below are listed the different data fields provided inside this dataset alongside a brief description:

| dateCrawled                    | name            | seller                                    | offerType           | price                               | abtest                                         | vehicleType      | yearOfRegistration                             | gearbox               | powerPS                    |
|--------------------------------|-----------------|-------------------------------------------|---------------------|-------------------------------------|------------------------------------------------|------------------|------------------------------------------------|-----------------------|----------------------------|
| When this ad was first crawled | Name of the car | Whether the seller is private or a dealer | The type of listing | The price on the ad to sell the car | Whether the listing is included in an A/B test | The vehicle Type | The year in which the car was first registered | The transmission type | The power of the car in PS |

| model              | odometer                              | monthOfRegistration                             | fuelType                       | brand                | notRepairedDamage                                 | dateCreated                                    | nrOfPictures                     | postalCode                                      | lastSeenOnline                           |
|--------------------|----------------------------------------|-------------------------------------------------|--------------------------------|----------------------|---------------------------------------------------|------------------------------------------------|----------------------------------|-------------------------------------------------|------------------------------------------|
| The car model name | How many kilometers the car has driven | The month in which the car was first registered | What type of fuel the car uses | The brand of the car | If the car has a damage which is not yet repaired | The date on which the eBay listing was created | The number of pictures in the ad | The postal code for the location of the vehicle | When the crawler saw this ad last online |

The aim of this project is to clean the data and analyze the included used car listings.

### Importing the libraries and reading the dataset

In [1]:
# Code #

import numpy as np
import pandas as pd

autos = pd.read_csv('autos.csv', encoding='Latin-1')

In [2]:
# Code (Rendering some values from the dataset into Jupyter)#

autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [3]:
# Code (Displaying some info from the dataset) #

autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

This far, we can see there are a total of 50,000 different entries organized in 20 columns. 5 of these columns are integer 64 type and the rest of the dataframe belongs to an object dtype. We have null values for the columns: vehicleType, gearbox, model, fuelType and notRepairedDamage. The full columns show a camelcase typo instead of Python's preferred snakecase.

Looking at this, the firs move will be to convert the column names from camelcase to snakecase and reword some of the column names based on the data dictionary to be more descriptive.

### Cleaning and managing the dataset labels

In [4]:
# Code (Managing dataset labels) #

columns_dataset = np.copy(autos.columns)

import re

def camel_to_snake(name):
  name = re.sub('(.)([A-Z][a-z]+)', r'\1_\2', name)
  return re.sub('([a-z0-9])([A-Z])', r'\1_\2', name).lower()

cleaned_columns = []

for c in columns_dataset:
    cleaned_c = camel_to_snake(c)
    cleaned_columns.append(cleaned_c)

cleaned_columns[7] = 'registration_year'
cleaned_columns[12] = 'registration_month'
cleaned_columns[15] = 'unrepaired_damage'
cleaned_columns[16] = 'ad_created'

#print(cleaned_columns)
autos.columns = cleaned_columns

In [5]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

Now let's do some basic data exploration to determine what other cleaning tasks need to be done. Initially we will look for:
- Text columns where all or almost all values are the same. These can often be dropped as they don't have useful information for analysis. 
- Examples of numeric data stored as text which can be cleaned and converted.

### Exploring the data and cleaning the dataframe

In [6]:
# Code (Exploring dataframe) #

autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-23 19:38:20  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

- Columns that have mostly one value that are candidates to be dropped. The columns seller and offer_type show only 2 unique values when they should contain more different values, these are candidates to be dropped from the dataframe. 
- Columns that need more investigation. 
- Numeric data stored as text that needs to be cleaned. For this purpose, the columns price and odometer are stored as object dtype (string) and both of them can be stored as numeric types.

In [7]:
# Code (Converting data columns to numeric type) #

autos.loc[:, 'odometer'] = autos.loc[:, 'odometer'].str.replace(',', '')
autos.loc[:, 'price'] = autos.loc[:, 'price'].str.replace(',', '')
autos.loc[:, 'odometer'] = autos.loc[:, 'odometer'].str.replace('km', '').astype(int)
autos.loc[:, 'price'] = autos.loc[:, 'price'].str.replace('$', '').astype(int)
autos.rename({'odometer':'odometer_km'}, axis=1, inplace=True)

At this point, we will look for outliers inside the data in the columns odometer_km and price. For this, we will analyze the columns using minimum and maximum values and look for any values that look unrealistically high or low that we might want to remove.

In [8]:
# Code (Looking for outliers) #

print('Odometer_km Column:\n')
print(autos.loc[:, 'odometer_km'].unique().shape) # To return the number of unique values
print(autos.loc[:, 'odometer_km'].describe()) # To return mean, std, min and max values
print(autos.loc[:, 'odometer_km'].value_counts().head(10)) # To return the count of the different values
print(autos.loc[:, 'odometer_km'].value_counts().sort_index(ascending=False).head(10)) # To return the highest or lowest values with their counts
print('Price Column:\n')
print(autos.loc[:, 'price'].unique().shape) # To return the number of unique values
print(autos.loc[:, 'price'].describe()) # To return mean, std, min and max values
print(autos.loc[:, 'price'].value_counts().head(10)) # To return the count of the different values
print(autos.loc[:, 'price'].value_counts().sort_index(ascending=False).head(15)) # To return the highest or lowest values with their counts

Odometer_km Column:

(13,)
count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64
150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
Name: odometer_km, dtype: int64
150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
40000       819
30000       789
Name: odometer_km, dtype: int64
Price Column:

(2357,)
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64
0       1421
500      781
1500     734
2500     643
1000     639
1200     639
600      531
800      498
3500     498
2000     460
Name: price, d

From the results, we can see that there is a big different between the odometer values for 150000 km and the rest of the values. Besides, there are no more higher entries, so we can infer that the values higher than 150000 are included inside this value. This scenario can undermine the whole dataset for this column if we are not careful with this. 

We can infer the same in the column price, where the value for 0 is the most common value observed. This can spoil the data if they are not excluded. The same apply for values that exceed 400000 dollars, they do not seem realistic and disturb the data.

In [9]:
# Code (Removing outliers) #

autos.loc[~autos['price'].between(50,400000), 'price'] = np.nan

In [10]:
# Code (Checking that outliers had been removed) #

print('Odometer_km Column:\n')
print(autos.loc[:, 'odometer_km'].unique().shape) # To return the number of unique values
print(autos.loc[:, 'odometer_km'].describe()) # To return mean, std, min and max values
print(autos.loc[:, 'odometer_km'].value_counts().head(10)) # To return the count of the different values
print(autos.loc[:, 'odometer_km'].value_counts().sort_index(ascending=False).head(10)) # To return the highest or lowest values with their counts
print('Price Column:\n')
print(autos.loc[:, 'price'].unique().shape) # To return the number of unique values
print(autos.loc[:, 'price'].describe()) # To return mean, std, min and max values
print(autos.loc[:, 'price'].value_counts().head(10)) # To return the count of the different values
print(autos.loc[:, 'price'].value_counts().sort_index(ascending=False).head(15)) # To return the highest or lowest values with their counts

Odometer_km Column:

(13,)
count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64
150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
Name: odometer_km, dtype: int64
150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
40000       819
30000       789
Name: odometer_km, dtype: int64
Price Column:

(2324,)
count     48347.000000
mean       5915.456099
std        9071.625794
min          50.000000
25%        1250.000000
50%        3000.000000
75%        7499.000000
max      350000.000000
Name: price, dtype: float64
500.0     781
1500.0    734
2500.0    643
1000.0    639
1200.0    639
600.0     531
3500.0    498
800.0     498
2000.0    460
999.0     

Let's now move on to the date columns and understand the date range the data covers.

There are 5 columns that should represent date values. Some of these columns were created by the crawler, some came from the website itself. We can differentiate by referring to the data dictionary:

- `date_crawled`: added by the crawler
- `last_seen`: added by the crawler
- `ad_created`: from the website
- `registration_month`: from the website
- `registration_year`: from the website

In [11]:
print(autos.dtypes)

date_crawled           object
name                   object
seller                 object
offer_type             object
price                 float64
abtest                 object
vehicle_type           object
registration_year       int64
gearbox                object
power_ps                int64
model                  object
odometer_km             int64
registration_month      int64
fuel_type              object
brand                  object
unrepaired_damage      object
ad_created             object
nr_of_pictures          int64
postal_code             int64
last_seen              object
dtype: object


As we can see, right now, the date_crawled, last_seen, and ad_created columns are all identified as string values by pandas. Because these three columns are represented as strings, we need to convert the data into a numerical representation so we can understand it quantitatively. 

The other two columns are represented as numeric values, so we can use methods like Series.describe() to understand the distribution without any extra data processing.

Let's first understand how the values in the three string columns are formatted. These columns all represent full timestamp values, like so:

In [12]:
print(autos[['date_crawled','ad_created','last_seen']][0:5])

          date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50


To understand the date range, we can extract just the date values, use Series.value_counts() to generate a distribution and then sort by the index.

In [13]:
# Code (Generating distribution for dates, including missing values)#

print('date_crawled Column:\n')
print(autos.loc[:, 'date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=True)) # Normalize to generate the distribution and sort_index to rank by date in ascending order
print('ad_created Column:\n')
print(autos.loc[:, 'ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=True))
print('last_seen Column:\n')
print(autos.loc[:, 'last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=True))

date_crawled Column:

2016-03-05    0.02538
2016-03-06    0.01394
2016-03-07    0.03596
2016-03-08    0.03330
2016-03-09    0.03322
2016-03-10    0.03212
2016-03-11    0.03248
2016-03-12    0.03678
2016-03-13    0.01556
2016-03-14    0.03662
2016-03-15    0.03398
2016-03-16    0.02950
2016-03-17    0.03152
2016-03-18    0.01306
2016-03-19    0.03490
2016-03-20    0.03782
2016-03-21    0.03752
2016-03-22    0.03294
2016-03-23    0.03238
2016-03-24    0.02910
2016-03-25    0.03174
2016-03-26    0.03248
2016-03-27    0.03104
2016-03-28    0.03484
2016-03-29    0.03418
2016-03-30    0.03362
2016-03-31    0.03192
2016-04-01    0.03380
2016-04-02    0.03540
2016-04-03    0.03868
2016-04-04    0.03652
2016-04-05    0.01310
2016-04-06    0.00318
2016-04-07    0.00142
Name: date_crawled, dtype: float64
ad_created Column:

2015-06-11    0.00002
2015-08-10    0.00002
2015-09-09    0.00002
2015-11-10    0.00002
2015-12-05    0.00002
2015-12-30    0.00002
2016-01-03    0.00002
2016-01-07    0.00002

Analyzing the results, we can see that for both the columns `date_crawled` and `last_seen` the dates move between March and April 2016. However, for `ad_created` we have values from 2015 or the first months of 2016. This could indicate that these values could be outliers values, that is to say, the advertiser who sold the car did not delete the post, for example.

Now, if we observe the `registration_year`column:

In [14]:
print(autos.loc[:, 'registration_year'].describe())

count    50000.000000
mean      2005.073280
std        105.712813
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64


Here, we can observe that minimum values of 1000 and maximum values of 9999. As this should be the year in which the car was first registered, these values are clearly wrong. Indeed, because a car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.



Let's count the number of listings with cars that fall outside the 1900 - 2016 interval and see if it's safe to remove those rows entirely, or if we need more custom logic.

In [15]:
autos.loc[~autos['registration_year'].between(1900,2016), 'registration_year'] = np.nan
print(autos.loc[:, 'registration_year'].value_counts(normalize=True)*100) # Distribution multiplied by 100 to show the result as percetages

2000.0    6.983426
2005.0    6.277588
1999.0    6.246356
2004.0    5.698759
2003.0    5.677938
2006.0    5.638378
2001.0    5.627967
2002.0    5.274007
1998.0    5.107437
2007.0    4.797202
2008.0    4.645207
2009.0    4.368285
1997.0    4.222537
2011.0    3.402182
2010.0    3.325144
1996.0    3.006579
2012.0    2.754643
2016.0    2.740068
1995.0    2.733822
2013.0    1.678188
2014.0    1.386691
1994.0    1.374198
1993.0    0.926543
2015.0    0.830765
1990.0    0.822437
1992.0    0.814108
1991.0    0.741234
1989.0    0.376863
1988.0    0.295661
1985.0    0.218622
            ...   
1977.0    0.045807
1966.0    0.045807
1975.0    0.039560
1969.0    0.039560
1965.0    0.035396
1964.0    0.024985
1910.0    0.018739
1963.0    0.018739
1959.0    0.014575
1961.0    0.012493
1956.0    0.010411
1962.0    0.008328
1958.0    0.008328
1937.0    0.008328
1950.0    0.006246
1934.0    0.004164
1941.0    0.004164
1951.0    0.004164
1954.0    0.004164
1955.0    0.004164
1957.0    0.004164
1952.0    0.

Observing the results, we can see that most of the cars are registered around the last 20 years (1985-2016).

When working with data on cars, it's natural to explore variations across different car brands. We can use aggregation to understand the `brand` column.

In [16]:
# Code (Using aggregation to explore the price by brand) #

brands = autos.loc[:, 'brand'].unique()
mean_price_by_brand = {}

for b in brands:
    selected_rows = autos.loc[autos.loc[:, 'brand'] == b, :]
    mean = selected_rows.loc[:, 'price'].mean()
    mean_price_by_brand[b] = mean

print('Mean price by brand sorted:\n', sorted(mean_price_by_brand.items(), key=lambda item: item[1], reverse=True))

# Second option (Aggregation in selected data) #

# brands = autos.loc[:, 'brand'].value_counts()
# selected_brands = brands[:20]

# mean_price_by_brand = {}

# for b in selected_brands:
#     selected_rows = autos.loc[autos.loc[:, 'brand'] == b, :]
#     mean = selected_rows.loc[:, 'price'].mean()
#     mean_price_by_brand[b] = mean

# print('Mean by brand sorted:\n', sorted(mean_price_by_brand.items(), key=lambda item: item[1], reverse=True))

Mean price by brand sorted:
 [('porsche', 46433.0), ('land_rover', 18934.272727272728), ('sonstige_autos', 12439.645652173913), ('jeep', 11573.638888888889), ('jaguar', 11525.554054054053), ('mini', 10566.824940047962), ('audi', 9252.86), ('mercedes_benz', 8571.013598100582), ('bmw', 8305.433473122379), ('chevrolet', 6644.226277372263), ('skoda', 6378.056628056628), ('dacia', 5897.736434108527), ('kia', 5889.298550724638), ('hyundai', 5382.935684647303), ('volkswagen', 5354.7044593413), ('toyota', 5148.0032733224225), ('volvo', 4911.680459770115), ('nissan', 4675.6945945945945), ('seat', 4339.289617486339), ('suzuki', 4152.322807017544), ('subaru', 4068.3861386138615), ('mazda', 4059.059539918809), ('alfa_romeo', 4054.471875), ('honda', 4010.4728682170544), ('citroen', 3772.8918128654973), ('ford', 3740.5440522100266), ('smart', 3533.2966714905933), ('chrysler', 3519.3846153846152), ('mitsubishi', 3406.286075949367), ('lancia', 3240.703703703704), ('saab', 3183.493670886076), ('peugeot

As per the results obtained, we can see an enormous differentiation between the mean price for certain _premium_ brands and the rest, with Porsche at the head of the pair price-brand.

For all the brands, let's use aggregation to understand the average mileage for those cars and if there's any visible link with mean price. To do this, we will create a new dataframe with just the values that we want to compare.

In [17]:
# Code (Using aggregation to explore the mileage by brand) #

mean_mileage_by_brand = {}

for b in brands:
    selected_rows = autos.loc[autos.loc[:, 'brand'] == b, :]
    mean = selected_rows.loc[:, 'odometer_km'].mean()
    mean_mileage_by_brand[b] = mean

print('Average mileage by brand sorted:\n', sorted(mean_mileage_by_brand.items(), key=lambda item: item[1], reverse=True))

Average mileage by brand sorted:
 [('saab', 143750.0), ('volvo', 138632.3851203501), ('rover', 136449.27536231885), ('chrysler', 133149.17127071825), ('bmw', 132521.64302818198), ('alfa_romeo', 131109.4224924012), ('mercedes_benz', 130886.14279678918), ('audi', 129643.9411627364), ('opel', 129298.66324848929), ('volkswagen', 128955.27276129878), ('renault', 128223.79367720465), ('peugeot', 127352.33516483517), ('jeep', 126409.09090909091), ('mitsubishi', 126293.10344827586), ('mazda', 125132.10039630119), ('subaru', 124449.54128440368), ('ford', 124131.93446392642), ('honda', 123709.27318295739), ('lancia', 123157.8947368421), ('seat', 122061.63655685441), ('daewoo', 121708.86075949368), ('jaguar', 121298.7012987013), ('citroen', 119764.62196861627), ('nissan', 118978.7798408488), ('land_rover', 118333.33333333333), ('fiat', 117037.4617737003), ('toyota', 115988.65478119935), ('daihatsu', 114843.75), ('kia', 112640.44943820225), ('skoda', 110947.83715012722), ('suzuki', 109334.47098976

In [18]:
# Code (Creating a dataframe from our dictionaries data) #

mpb_series = pd.Series(mean_price_by_brand)
mmb_series = pd.Series(mean_mileage_by_brand)
df = pd.DataFrame(mpb_series, columns=['mean_price'])
df.loc[:, 'avg_mileage_km'] = mmb_series
print(df)

                  mean_price  avg_mileage_km
alfa_romeo       4054.471875   131109.422492
audi             9252.860000   129643.941163
bmw              8305.433473   132521.643028
chevrolet        6644.226277    99522.968198
chrysler         3519.384615   133149.171271
citroen          3772.891813   119764.621969
dacia            5897.736434    84728.682171
daewoo           1093.600000   121708.860759
daihatsu         1628.426230   114843.750000
fiat             2798.303175   117037.461774
ford             3740.544052   124131.934464
honda            4010.472868   123709.273183
hyundai          5382.935685   106782.786885
jaguar          11525.554054   121298.701299
jeep            11573.638889   126409.090909
kia              5889.298551   112640.449438
lada             2647.724138    86774.193548
lancia           3240.703704   123157.894737
land_rover      18934.272727   118333.333333
mazda            4059.059540   125132.100396
mercedes_benz    8571.013598   130886.142797
mini      

There are no significant differences between the average mileage for the brands. This could be impacted by the fact that the values for 150000+ km are included inside the value of 150000. This definitely alters the outputs received.

This is the fundamental reason by which is so important to deal with outliers carefully. From this, we can remake the treatment of the outliers to generate a new analysis and conclusions if required. For example, a good aproximation would be to remove all the rows that contain a value for the mileage of 150000, as we have already seen they alter the results. This could be done assigning a NaN value for all the 150000 values followed by a Dataframe.dropna() method to delete the impacted rows: 
`autos.dropna(subset=['odometer_km'])`.

Another steps to consider:

1. Data cleaning:
    - Identify categorical data that uses german words, translate them and map the values to their english counterparts.
    - Convert the dates to be uniform numeric data, so "2016-03-21" becomes the integer 20160321.
    - See if there are particular keywords in the name column that you can extract as new columns. For example, any particular model of a brand.

2. Data analysis:
    - Find the most common brand/model combinations.
    - Split the `odometer_km` into groups, and use aggregation to see if average prices follows any patterns based on the mileage.
    - How much cheaper are cars with damage than their non-damaged counterparts?